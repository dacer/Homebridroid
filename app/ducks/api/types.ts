const LOGIN = "homebridroid/api/LOGIN"
const GET_DEIVCES = "homebridroid/api/GET_DEIVCES"

export {
    LOGIN,
    GET_DEIVCES
};