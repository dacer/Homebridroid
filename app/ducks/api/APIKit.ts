import axios, { AxiosInstance } from 'axios';

class APIKit {
  private static instance: AxiosInstance;

  /**
   * The Singleton's constructor should always be private to prevent direct
   * construction calls with the `new` operator.
   */
  private constructor() { }

  /**
   * The static method that controls the access to the singleton instance.
   *
   * This implementation let you subclass the Singleton class while keeping
   * just one instance of each subclass around.
   */
  public static getInstance(baseUrl?: string, token?: string): AxiosInstance {
      if (!APIKit.instance) {
          APIKit.instance = axios.create({
            baseURL: baseUrl,
          });
      }

      this.setClientToken(token)
      return APIKit.instance;
  }

  private static setClientToken = (token?: string) => {
    APIKit.instance.interceptors.request.use(function(config) {
      if (token) config.headers["Authorization"] = `bearer ${token}`
      config.headers["accept"] = "application/json, text/plain, */*"
      config.headers["Content-Type"] = "application/json"
      return config;
    })
  }
}

export default APIKit;