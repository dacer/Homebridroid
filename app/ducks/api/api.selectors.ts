import { pathOr } from "ramda";

import { DeviceLayout, namespace } from "./api.slice";

const getToken = (state: any) =>
  pathOr("", [namespace, "token"], state);

const getDevices = (state: any) =>
  pathOr([], [namespace, "devices"], state) as DeviceLayout[];

export default {
  getToken, getDevices
};
