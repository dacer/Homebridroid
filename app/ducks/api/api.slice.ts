import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AuthInfo } from "../userPreferences/userPreferences.slice";
import APIKit from "./APIKit";
import { GET_DEIVCES, LOGIN } from "./types";

export const login = createAsyncThunk(LOGIN, async (
  {username, password, serverAddress} : AuthInfo
) => {
  if (serverAddress.endsWith('/')) {
    serverAddress = serverAddress.slice(0, -1)
  }
  const response = await APIKit.getInstance(serverAddress).post('/api/auth/login', {
    username,
    password
  })
  try {
    const token = (await response.data).access_token as string
    return token
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
})

export const getDevices = createAsyncThunk(GET_DEIVCES, async (
  {serverAddress, token} : AuthInfo
) => {
  try {
    const devices = (await APIKit.getInstance(serverAddress, token).get('/api/accessories')).data as Device[]
    const layoutList = (await APIKit.getInstance(serverAddress, token).get('/api/accessories/layout')).data as Layout[]

    const deviceLayoutList: DeviceLayout[] = layoutList.map(layout => (
      {
        name: layout.name, 
        services: layout.services.map(s => ({
          uniqueId: s.uniqueId,
          serviceName: devices.find(d => d.uniqueId === s.uniqueId)!.serviceName
        }))
      }
    ))
    console.log(`deviceLayoutList: ${JSON.stringify(deviceLayoutList)}`)
    return deviceLayoutList
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
})

const initialState = {
  token: "",
  devices: [],
  loading: 'idle'
} as ApiState

interface ApiState {
  token: string,
  devices: DeviceLayout[]
  loading: 'idle' | 'pending' | 'succeeded' | 'failed'
}

export type Device = {
  uniqueId: string,
  serviceName: string
}

type Layout = {
  name: string,
  services: {
    uniqueId: string,
  }[]
}

export type DeviceLayout = {
  name: string,
  services: {
    uniqueId: string,
    serviceName: string
  }[]
}

const emissions = createSlice({
  name: "api",
  initialState,
  reducers: {
    logout: (state, action: PayloadAction<typeof undefined>) => {
      state.token = ""
    }
  },
  extraReducers: builder => {
    builder
      .addCase(login.pending, (state, action) => {
        state.devices = []
      })
      .addCase(login.fulfilled, (state, { payload }) => {
        state.loading = 'idle'
        state.token = payload
        console.log(`${payload} <- payload`)
      })
      .addCase(getDevices.fulfilled, (state, { payload }) => {
        state.loading = 'idle'
        state.devices = payload
      })
      
    }
});

const { logout } = emissions.actions;

export const actions = { login, getDevices, logout };

export const namespace = emissions.name;

export const reducer = emissions.reducer;
