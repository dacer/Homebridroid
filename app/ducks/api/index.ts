import { actions, namespace, reducer } from "./api.slice";
import selectors from "./api.selectors";

export default { actions, namespace, reducer, selectors };
