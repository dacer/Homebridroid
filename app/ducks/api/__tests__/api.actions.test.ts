import api from "..";
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe("Api actions should", () => {
  it("export expected actions", () => expect(api.actions).toMatchSnapshot());

  it("be able to logout", () => {
    const expectedAction = {
      type: api.actions.logout.toString(),
      payload: undefined,
    };

    expect(
      api.actions.logout()
    ).toEqual(expectedAction);
  });
});
