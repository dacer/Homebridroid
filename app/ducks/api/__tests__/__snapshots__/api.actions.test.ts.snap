// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api actions should export expected actions 1`] = `
Object {
  "getDevices": [Function],
  "login": [Function],
  "logout": [Function],
}
`;
